runtime/runtime-common/src/precompiles/zk_verify.rs

```diff
@@ -0,0 +1,183 @@
+
+		if input_stripped.len() < 704 {
+			return Err(PrecompileFailure::from(ExitError::InvalidRange));
+		}
+
+		let proof_a_x = &input_stripped[i256_r.next()];
+		let proof_a_y = &input_stripped[i256_r.next()];
+		let proof_b_x1 = &input_stripped[i256_r.next()];
+		let proof_b_x2 = &input_stripped[i256_r.next()];
+		let proof_b_y1 = &input_stripped[i256_r.next()];
+		let proof_b_y2 = &input_stripped[i256_r.next()];
+		let proof_c_x = &input_stripped[i256_r.next()];
+		let proof_c_y = &input_stripped[i256_r.next()];
+
+		let vk_alpha_x = &input_stripped[i256_r.next()];
+		let vk_alpha_y = &input_stripped[i256_r.next()];
+		let vk_beta_x1 = &input_stripped[i256_r.next()];
+		let vk_beta_x2 = &input_stripped[i256_r.next()];
+		let vk_beta_y1 = &input_stripped[i256_r.next()];
+		let vk_beta_y2 = &input_stripped[i256_r.next()];
+		let vk_gamma_x1 = &input_stripped[i256_r.next()];
+		let vk_gamma_x2 = &input_stripped[i256_r.next()];
+		let vk_gamma_y1 = &input_stripped[i256_r.next()];
+		let vk_gamma_y2 = &input_stripped[i256_r.next()];
+		let vk_delta_x1 = &input_stripped[i256_r.next()];
+		let vk_delta_x2 = &input_stripped[i256_r.next()];
+		let vk_delta_y1 = &input_stripped[i256_r.next()];
+		let vk_delta_y2 = &input_stripped[i256_r.next()];
+
+		// It seems like whole data should be decoded to decode arrays.
+		let types = &[
+			ethabi::ParamType::Uint(256), // proof a x
+			ethabi::ParamType::Uint(256), // proof a y
+			ethabi::ParamType::Uint(256), // proof b x1
+			ethabi::ParamType::Uint(256), // proof b x2
+			ethabi::ParamType::Uint(256), // proof b y1
+			ethabi::ParamType::Uint(256), // proof b y2
+			ethabi::ParamType::Uint(256), // proof c x
+			ethabi::ParamType::Uint(256), // proof c y
+			ethabi::ParamType::Uint(256), // vk alpha x
+			ethabi::ParamType::Uint(256), // vk alpha y
+			ethabi::ParamType::Uint(256), // vk beta x1
+			ethabi::ParamType::Uint(256), // vk beta x2
+			ethabi::ParamType::Uint(256), // vk beta y1
+			ethabi::ParamType::Uint(256), // vk beta y2
+			ethabi::ParamType::Uint(256), // vk gamma x1
+			ethabi::ParamType::Uint(256), // vk gamma x2
+			ethabi::ParamType::Uint(256), // vk gamma y1
+			ethabi::ParamType::Uint(256), // vk gamma y2
+			ethabi::ParamType::Uint(256), // vk delta x1
+			ethabi::ParamType::Uint(256), // vk delta x2
+			ethabi::ParamType::Uint(256), // vk delta y1
+			ethabi::ParamType::Uint(256), // vk delta y2
+			ethabi::ParamType::Array(Box::from(ethabi::ParamType::FixedArray(
+				Box::from(ethabi::ParamType::Uint(256)),
+				2,
+			))), // vk ic
+			ethabi::ParamType::Array(Box::from(ethabi::ParamType::Uint(256))), // input
+		];
+		let decoded = match ethabi::decode(types, input_stripped) {
+			Ok(v) => v,
+			Err(_e) => {
+				return Err(PrecompileFailure::from(ExitError::InvalidRange));
+			}
+		};
+
+		let pub_arr = match decoded[decoded.len() - 1].clone().into_array() {
+			Some(v) => v,
+			None => return Err(PrecompileFailure::from(ExitError::InvalidRange)),
+		};
```

Comment:
We can use [map_err](https://doc.rust-lang.org/stable/std/result/enum.Result.html#method.map_err) here.